les/costtracker.go

```diff
@@ -193,26 +232,47 @@ func (ct *costTracker) gfLoop() {
 	}
 	gf := gfSum / gfWeight
 	ct.gf = gf
+	totalRecharge := ct.utilTarget * gf
 	ct.gfUpdateCh = make(chan gfUpdate, 100)
 
 	go func() {
+		saveCostFactor := func() {
+			var data [16]byte
+			binary.BigEndian.PutUint64(data[0:8], math.Float64bits(gfSum))
+			binary.BigEndian.PutUint64(data[8:16], math.Float64bits(gfWeight))
+			ct.db.Put([]byte(gfDbKey), data[:])
+			log.Debug("global cost factor saved", "sum", time.Duration(gfSum), "weight", time.Duration(gfWeight))
+		}
+		saveTicker := time.NewTicker(time.Minute * 10)
+
 		for {
 			select {
 			case r := <-ct.gfUpdateCh:
 				now := mclock.Now()
 				max := r.servingTime * gf
+				if ct.logRelCost != nil && r.avgTime > 1e-20 {
+					ct.logRelCost.Update(max / r.avgTime)
+				}
+				if r.servingTime > 1000000000 {
+					ct.logger.Event(fmt.Sprintf("Very long servingTime = %f  avgTime = %f  costFactor = %f", r.servingTime, r.avgTime, gf))
+				}
+				if max > r.avgTime*maxCostFactor {
```

Comment:
I did a few experiments here. I think we should remove the `recap` here.

Basically what we want to do is: accumulate recent usage, when it reaches the threshold, re-calculate the `factor`.
But we shouldn't recap the `max`. For example in my server the `factor` is 30, while the `relativeCost` is about 40. So that we always try to recap the max to a smaller value, then the threshold can never be reached.

Reply:
Btw the ideal `relativeCost` should be around 1.

Reply:
I see the problem, the control mechanism can get stuck once the factor goes very high. I will think about it a bit more before fixing it but I will try to do it today.